internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,203 @@
+		}
+	}
+
+	// For other services: forwarders will be started in ServiceStarted
+
+	// Update buffers and iterators
+	m.buffers = buffers
+	m.iterators = iterators
+
+	// Update services
+	m.services = map[string]*plan.Service{}
+	for serviceName, service := range pl.Services {
+		m.services[serviceName] = service.Copy()
+	}
+}
+
+// selectTargets goes through the given plan and determines which service's
+// logs should be forwarded to which log targets.
+func selectTargets(p *plan.Plan) map[string]map[string]*plan.LogTarget {
+	planTargets := make(map[string]map[string]*plan.LogTarget, len(p.Services))
+
+	for serviceName, service := range p.Services {
+		planTargets[serviceName] = make(map[string]*plan.LogTarget, len(service.LogTargets))
+
+		for _, targetName := range service.LogTargets {
+			target, ok := p.LogTargets[targetName]
+			if !ok {
+				logger.Noticef("Internal error: log manager cannot find target %q (specified by service %q in plan)",
+					targetName, serviceName)
+				continue
+			}
+			if target.Selection != plan.DisabledSelection {
+				planTargets[serviceName][targetName] = target
+			}
+		}
+
+		// Add opt-out targets if none specified
+		if len(service.LogTargets) == 0 {
+			for targetName, target := range p.LogTargets {
+				// Default (unset) is opt-out
+				if target.Selection == plan.OptOutSelection || target.Selection == plan.UnsetSelection {
+					planTargets[serviceName][targetName] = target
+				}
+			}
+		}
+	}
+
+	return planTargets
+}
+
+// ServiceStarted notifies the log manager that the named service has started,
+// and provides a reference to the service's log buffer.
+func (m *LogManager) ServiceStarted(serviceName string, buffer *servicelog.RingBuffer) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.buffers[serviceName] = buffer
+	m.iterators[serviceName] = make(map[plan.LogTarget]servicelog.Iterator,
+		len(m.planTargets[serviceName]))
+
+	for _, target := range m.planTargets[serviceName] {
+		iterator := m.buffers[serviceName].HeadIterator(0)
+		m.iterators[serviceName][*target] = iterator
+
+		m.startForwarder(serviceName, target, iterator)
+	}
+}
+
+// startForwarder starts a new forwarder for the specified service and target,
+// and adds this forwarder to the LogManager's list.
+func (m *LogManager) startForwarder(
+	serviceName string, target *plan.LogTarget, iterator servicelog.Iterator,
+) {
+	f, err := newLogForwarder(serviceName, target, iterator)
+	if err != nil {
+		logger.Noticef("Internal error: cannot create forwarder %q -> %q: %v",
+			serviceName, target.Name, err)
+		return
+	}
+
+	m.forwarders = append(m.forwarders, f)
+	logger.Debugf("Started forwarding logs for service %q to target %q",
+		serviceName, target.Name)
+	go f.forward()
+}
+
+// Ensure implements overlord.StateManager.
+func (m *LogManager) Ensure() error {
+	return nil
+}
+
+// Stop implements overlord.StateStopper and stops all log forwarding.
+func (m *LogManager) Stop() {
+	wg := sync.WaitGroup{}
```

Comment:
Nice idea to call all the `f.stop()`s concurrently!

Reply:
I was thinking of doing something similar when I stop and restart them in `planChanged`. Now that I think of that, it might even be worth having a tomb for each forwarder.